src/diffusers/pipelines/faster_cache_utils.py

```diff
@@ -0,0 +1,305 @@
+# Copyright 2024 The HuggingFace Team. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from dataclasses import dataclass
+from typing import Any, Callable, Optional, Tuple
+
+import torch
+import torch.fft as FFT
+import torch.nn as nn
+
+from ..models.attention_processor import Attention
+from ..models.hooks import ModelHook, add_hook_to_module
+from ..utils import logging
+from .pipeline_utils import DiffusionPipeline
+
+
+logger = logging.get_logger(__name__)  # pylint: disable=invalid-name
+
+
+_ATTENTION_CLASSES = (Attention,)
+
+_SPATIAL_ATTENTION_BLOCK_IDENTIFIERS = ("blocks", "transformer_blocks")
+_TEMPORAL_ATTENTION_BLOCK_IDENTIFIERS = ("temporal_transformer_blocks",)
+
+
+@dataclass
+class FasterCacheConfig:
```

Comment:
For future.

Once we start to have more than one caching mechanism implemented, we could think about about a generic `caching_utils.py` and have all the other caching utilities subclass from it.

Reply:
Would like to note that the store-and-reuse of tensors itself (cache), and how we do it, is orthogonal to the algorithmic implementations itself. For this reason, would prefer to:
- First introduce PAB, FasterCache, AdaCache, TeaCache
- See what are common and then perform refactors to design
- Until refactor, let the output cache be stored simply as a device tensor in `self.cache` for each hook directly
- Once we have a sense of what could be re-used and if it makes sense to have cache-related offloading, by profiling extra memory requirements and extra time overhead from offload, only then move to adding any caching/offloading utilities.

Reply:
Yes this makes complete sense.

Reply:
But I think the common things could be factored out later independent of offloading, no?